Add check if the gift wrapping is disabled

marsaldev opened the pull request:
Don't show the line in the order summary. I've added a reminder hint in the BO form

<!-----------------------------------------------------------------------------
Thank you for contributing to the PrestaShop project! 

Please take the time to edit the "Answers" rows below with the necessary information.

Check out our contribution guidelines to find out how to complete it:
https://devdocs.prestashop.com/1.7/contribute/contribution-guidelines/#pull-requests
------------------------------------------------------------------------------>

| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | develop
| Description?      | If you disable the Gift Option, in the order summary the gift wrapping line is still present. It will be confusing.
| Type?             | bug fix
| Category?         | FO
| BC breaks?        | no
| Deprecations?     | no
| Fixed ticket?     | Fixes #22825.
| How to test?      | Apply edits, regenerate e-mails.
| Possible impacts? | If the merchant enable or disable _Gift Wrapping_ option he/she needs to re-generate e-mails template. In this PR I've added a small hint in the BO as help text near the option

If this PR will be approved, I will made a PR in https://github.com/PrestaShop/mjml-theme-converter

<!-- Click the form's "Preview" button to make sure the table is functional in GitHub. Thank you! -->

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/prestashop/prestashop/27890)
<!-- Reviewable:end -->

Commit:
Add check if the gift wrapping is disabled

Don't show the line in the order summary. I've added a reminder hint in the BO form

Commit:
fix static call

matks commented on mails/themes/modern/core/order_conf.html.twig:
```diff
@@ -178,18 +178,18 @@
                                       <th bgcolor="#FDFDFD">{{ 'Total price'|trans({}, 'Emails.Body', locale)|raw }}</th>
                                     </tr>
-                                    {% if templateType == 'html' %}
-{products}
```
I think the indentation is now broken. Can we fix it?

marsaldev replied:
which is the correct one? xD
I think the proposed one, although this file is generated by mjml 🤔 wdyt?

matks replied:
Mmmm ... interesting.
Do you think the fact there is blank space or not has an impact on how Mail clients would render the mail?

marsaldev replied:
🤔 I don't think so, now we follow the table indentation, in theory there should be no issues on rendering the html generated

matks commented on src/Adapter/MailTemplate/MailTemplateTwigRenderer.php:
```diff
@@ -67,12 +71,14 @@ class MailTemplateTwigRenderer implements MailTemplateRendererInterface
         LayoutVariablesBuilderInterface $variablesBuilder,
-        HookDispatcherInterface $hookDispatcher
+        HookDispatcherInterface $hookDispatcher,
+        ConfigurationInterface $configuration
```
This is a backward compatibility break because if someone is currently using this constructor, after your PR is merged it will now crash (fatal error) because the new constructor argument will be missing in his code.

But good news: next version is a major version! BC breaks are allowed 😄  . However we need to list them in the PR description.

matks replied:
By the way PHPStan (see below) reported issues _exactly_ because of this problem: in some unit tests, MailTemplateTwigRenderer is being constructed. You did not update the code of these tests so these tests now crash.

marsaldev replied:
Thanks @matks! 
One question, I need to write the unit tests too?

matks replied:
My opinion is that there is not a lot of value to be tested here. You give `$configuration` to the class, then it uses `get()` to retrieve `giftWrapping` and then we expose it as a smarty variable. We do not transform the data, we just transfer it.

marsaldev replied:
Got it, but I need to update the code in the existing unit tests, right?

matks replied:
Yes. Existing unit tests need to be updated so that they pass

matks reviewed (changes requested):
Nice feature @marsaldev !

I reported a few issues, quite easy to fix. Also PHP CS Fixer is complaining (ref https://github.com/PrestaShop/PrestaShop/runs/5510708553?check_suite_focus=true). You can make him happy by running `php vendor/bin/php-cs-fixer fix`

Commit:
fix syntax with php-cs-fixer

Commit:
updated unit test

Julievrz commented on src/PrestaShopBundle/Form/Admin/Configure/ShopParameters/OrderPreferences/GiftOptionsType.php:
```diff
@@ -77,7 +77,7 @@ public function buildForm(FormBuilderInterface $builder, array $options)
                 'required' => false,
                 'label' => $this->trans('Offer gift wrapping', 'Admin.Shopparameters.Feature'),
-                'help' => $this->trans('Suggest gift-wrapping to customers.', 'Admin.Shopparameters.Help'),
+                'help' => $this->trans('Suggest gift-wrapping to customers. Remember to regenerate e-mails template if you activate or deactivate this option', 'Admin.Shopparameters.Help'),
```
```suggestion
                'help' => $this->trans(Remember to regenerate email templates after enabling or disabling this feature.', 'Admin.Shopparameters.Help'),
```

I removed the first sentence as it was a repetition of the label. Also, maybe we could specify which templates should be regenerated?

marsaldev replied:
As far as I remember you can't chose a specific one, the regeneration process will regenerate all the e-mails (Design > Email theme)
Maybe we could add the path of the BO, like:
`Remember to regenerate email templates after enabling or disabling this option in Design > Email theme`

Julievrz replied:
Good idea, then why not add a link? :-)

`Remember to regenerate email templates after enabling or disabling this option in [1]Design > Email theme[\1].`

marsaldev replied:
Done 👍🏻

Commit:
Improved help label with link to Email Theme page

This require the integration of the Router service, the constructor and the yml file have been updated

Progi1984 commented on src/PrestaShopBundle/Resources/config/services/adapter/mail_template.yml:
```diff
@@ -8,6 +8,7 @@ services:
       - '@twig'
       - '@prestashop.core.mail_template.variables_builder'
       - '@prestashop.core.hook.dispatcher'
+      - '@prestashop.adapter.legacy.configuration'
```
```suggestion
      - "@=service('prestashop.adapter.legacy.configuration').getBoolean('PS_GIFT_WRAPPING')"
```

Progi1984 commented on src/Adapter/MailTemplate/MailTemplateTwigRenderer.php:
```diff
@@ -57,6 +58,9 @@ class MailTemplateTwigRenderer implements MailTemplateRendererInterface
     /** @var TransformationCollection */
     private $transformations;
 
+    /** @var ConfigurationInterface */
+    private $configuration;
```
```suggestion
    /** @var bool */
    private $hasGiftWrapping;
```

Progi1984 commented on src/Adapter/MailTemplate/MailTemplateTwigRenderer.php:
```diff
@@ -67,12 +71,14 @@ class MailTemplateTwigRenderer implements MailTemplateRendererInterface
         LayoutVariablesBuilderInterface $variablesBuilder,
-        HookDispatcherInterface $hookDispatcher
+        HookDispatcherInterface $hookDispatcher,
+        ConfigurationInterface $configuration
```
```suggestion
        bool $hasGiftWrapping
```

Progi1984 commented on src/Adapter/MailTemplate/MailTemplateTwigRenderer.php:
```diff
@@ -67,12 +71,14 @@ class MailTemplateTwigRenderer implements MailTemplateRendererInterface
         $this->variablesBuilder = $variablesBuilder;
         $this->hookDispatcher = $hookDispatcher;
         $this->transformations = new TransformationCollection();
+        $this->configuration = $configuration;
```
```suggestion
        $this->hasGiftWrapping = $hasGiftWrapping;
```

Progi1984 commented on src/Adapter/MailTemplate/MailTemplateTwigRenderer.php:
```diff
@@ -122,6 +128,7 @@ private function render(
     ) {
         $layoutVariables = $this->variablesBuilder->buildVariables($layout, $language);
         $layoutVariables['templateType'] = $templateType;
+        $layoutVariables['giftWrapping'] = $this->configuration->get('PS_GIFT_WRAPPING');
```
```suggestion
        $layoutVariables['giftWrapping'] = $this->hasGiftWrapping;
```

Progi1984 reviewed (changes requested):
@marsaldev Thank you for your contribution. Could you document the BC Break, please ?

Julievrz commented on src/PrestaShopBundle/Form/Admin/Configure/ShopParameters/OrderPreferences/GiftOptionsType.php:
```diff
@@ -77,7 +92,13 @@ public function buildForm(FormBuilderInterface $builder, array $options)
                 'required' => false,
                 'label' => $this->trans('Offer gift wrapping', 'Admin.Shopparameters.Feature'),
-                'help' => $this->trans('Suggest gift-wrapping to customers.', 'Admin.Shopparameters.Help'),
+                'help' => $this->trans('Remember to regenerate email templates after enabling or disabling this option in [1]Design > Email theme[/1].',
```
```suggestion
                'help' => $this->trans('Remember to regenerate email templates in [1]Design > Email theme[/1] after enabling or disabling this feature.',
```

Otherwise, it looks like the feature can be enabled/disabled in Design>Email theme

marsaldev committed:
Apply suggestions from code review

Co-authored-by: Progi1984 <progi1984@gmail.com>
Co-authored-by: Julie Varisellaz <70583503+Julievrz@users.noreply.github.com>

marsaldev commented:
> @marsaldev Thank you for your contribution. Could you document the BC Break, please ?

Yes no problem, but where 😊 ?

~ ~ ~

I've accepted the requested changes, but I don't understand the real difference to provide the single value from the yml file rather than the solution proposed, give the Configuration to the constructor and then retrieve the value. 

I think it could be more useful to have the Configuration element available at class level also in the future, because if in the future we want add another setting we have to update the yml, the constructor and the tests as well. This will introduce another BC Break. It makes sense?
I don't think so 😊

Progi1984 commented on src/PrestaShopBundle/Form/Admin/Configure/ShopParameters/OrderPreferences/GiftOptionsType.php:
```diff
@@ -51,16 +52,30 @@ class GiftOptionsType extends TranslatorAwareType
+    /**
+     * @param TranslatorInterface $translator
+     * @param array $locales
+     * @param $defaultCurrencyIsoCode
```
```suggestion
     * @param string $defaultCurrencyIsoCode
```

Progi1984 commented on tests/Unit/Adapter/MailTemplate/MailTemplateTwigRendererTest.php:
```diff
@@ -60,7 +61,12 @@ public function testConstructor()
         ;
 
-        $generator = new MailTemplateTwigRenderer($engineMock, $builderMock, $dispatcherMock);
+        $configurationMock = $this->getMockBuilder(ConfigurationInterface::class)
+            ->disableOriginalConstructor()
+            ->getMock()
+        ;
+
```
```suggestion
```

Progi1984 commented on tests/Unit/Adapter/MailTemplate/MailTemplateTwigRendererTest.php:
```diff
@@ -60,7 +61,12 @@ public function testConstructor()
+            ->getMock()
+        ;
+
+        $generator = new MailTemplateTwigRenderer($engineMock, $builderMock, $dispatcherMock, $configurationMock);
```
```suggestion
        $generator = new MailTemplateTwigRenderer($engineMock, $builderMock, $dispatcherMock, false);
```

Progi1984 commented on tests/Unit/Adapter/MailTemplate/MailTemplateTwigRendererTest.php:
```diff
@@ -88,12 +94,18 @@ public function testFileNotFound()
             ->disableOriginalConstructor()
             ->getMock()
         ;
+        /** @var ConfigurationInterface $configurationMock */
+        $configurationMock = $this->getMockBuilder(ConfigurationInterface::class)
+            ->disableOriginalConstructor()
+            ->getMock()
+        ;
```
```suggestion
```

Progi1984 commented on tests/Unit/Adapter/MailTemplate/MailTemplateTwigRendererTest.php:
```diff
@@ -88,12 +94,18 @@ public function testFileNotFound()
             $this->createVariablesBuilderMock($expectedVariables, $expectedLanguage),
-            $dispatcherMock
+            $dispatcherMock,
+            $configurationMock
```
```suggestion
            false
```

Progi1984 commented on tests/Unit/Adapter/MailTemplate/MailTemplateTwigRendererTest.php:
```diff
@@ -378,4 +395,21 @@ private function createMailLayoutMock(array $expectedPaths)
 
         return $mailLayoutMock;
     }
+
+    /**
+     * @return MockObject|ConfigurationInterface
+     */
+    private function createConfigurationMock()
+    {
+        $configurationMock = $this->getMockBuilder(ConfigurationInterface::class)
+            ->disableOriginalConstructor()
+            ->getMock()
+        ;
+
+        $configurationMock
+            ->method('get')
+            ->willReturn(1);
+
+        return $configurationMock;
+    }
```
```suggestion
```

Progi1984 commented on tests/Unit/Adapter/MailTemplate/MailTemplateTwigRendererTest.php:
```diff
@@ -193,14 +209,15 @@ public function testRenderWithTransformations()
             $this->createVariablesBuilderMock($expectedVariables, $expectedLanguage),
-            $this->createHookDispatcherMock($mailLayout, MailTemplateInterface::HTML_TYPE, 1)
+            $this->createHookDispatcherMock($mailLayout, MailTemplateInterface::HTML_TYPE, 1),
+            $this->createConfigurationMock()
```
```suggestion
            true
```

Progi1984 commented on tests/Unit/Adapter/MailTemplate/MailTemplateTwigRendererTest.php:
```diff
@@ -171,14 +186,15 @@ public function testRenderTxtFallback()
             $this->createVariablesBuilderMock($expectedVariables, $expectedLanguage),
-            $this->createHookDispatcherMock($mailLayout, MailTemplateInterface::TXT_TYPE)
+            $this->createHookDispatcherMock($mailLayout, MailTemplateInterface::TXT_TYPE),
+            $this->createConfigurationMock()
```
```suggestion
            true
```

Progi1984 commented on tests/Unit/Adapter/MailTemplate/MailTemplateTwigRendererTest.php:
```diff
@@ -149,14 +163,15 @@ public function testRenderTxt(): void
             $this->createVariablesBuilderMock($expectedVariables, $expectedLanguage),
-            $this->createHookDispatcherMock($mailLayout, MailTemplateInterface::TXT_TYPE)
+            $this->createHookDispatcherMock($mailLayout, MailTemplateInterface::TXT_TYPE),
+            $this->createConfigurationMock()
```
```suggestion
            true
```

Progi1984 commented on tests/Unit/Adapter/MailTemplate/MailTemplateTwigRendererTest.php:
```diff
@@ -128,14 +141,15 @@ public function testRenderHtmlWithFallback(): void
             $this->createVariablesBuilderMock($expectedVariables, $expectedLanguage),
-            $this->createHookDispatcherMock($mailLayout, MailTemplateInterface::HTML_TYPE)
+            $this->createHookDispatcherMock($mailLayout, MailTemplateInterface::HTML_TYPE),
+            $this->createConfigurationMock()
```
```suggestion
            true
```

Progi1984 commented on tests/Unit/Adapter/MailTemplate/MailTemplateTwigRendererTest.php:
```diff
@@ -106,14 +118,15 @@ public function testRenderHtml(): void
             $this->createVariablesBuilderMock($expectedVariables, $expectedLanguage),
-            $this->createHookDispatcherMock($mailLayout, MailTemplateInterface::HTML_TYPE)
+            $this->createHookDispatcherMock($mailLayout, MailTemplateInterface::HTML_TYPE),
+            $this->createConfigurationMock()
```
```suggestion
            true
```

Progi1984 reviewed (dismissed):
@marsaldev Thank you for your contribution. I have brought you some comments about your test for fixing them.
Could you document the BC Break and add a linked PR on the devdocs : https://devdocs.prestashop.com/8/modules/core-updates/8.0/ ?

marsaldev committed:
Apply suggestions from code review

Co-authored-by: Progi1984 <progi1984@gmail.com>

Commit:
Fix for PHPCSFixer

marsaldev commented:
> @marsaldev Thank you for your contribution. I have brought you some comments about your test for fixing them. Could you document the BC Break and add a linked PR on the devdocs : https://devdocs.prestashop.com/8/modules/core-updates/8.0/ ?

Created a PR for the devdocs 😉 
Applied your suggestions + applied phpcsfixer suggestion

HanaRebaiQA reviewed (commented):
Hello @marsaldev 

I have checked this PR. The issue is not fixed.

1. Disable the option `Offer gift wrapping` from BO > Shop Parameters > Order settings > Gift options
2. Regenerate email template from Design > Email Theme (With Modern and classic theme)
3. Create an order from BO 
4. Check the order summary of the **new order** received in the email
5. The line `Gift-wrapping` still exist 

Here is the example : 
![image](https://user-images.githubusercontent.com/92912932/162752254-10fadba3-7e21-4de0-9f89-66f02003c4ab.png)
![image](https://user-images.githubusercontent.com/92912932/162752279-9aa8a61e-a65d-4148-b040-9da202eff1b0.png)
![image](https://user-images.githubusercontent.com/92912932/162752294-86b8f3d7-8f79-4bcf-a414-a9accdec2f8c.png)
![image](https://user-images.githubusercontent.com/92912932/162752907-097c27f6-22c6-4397-a16a-7abc2c108444.png)

https://user-images.githubusercontent.com/92912932/162754317-37f75a3c-3ab5-4f85-a73b-195378b98de4.mp4


The same for modern or classic theme 
![image](https://user-images.githubusercontent.com/92912932/162754075-5cab64a1-9ce0-4b8c-8928-6a590098204b.png)

https://user-images.githubusercontent.com/92912932/162753493-66f7720a-a3a6-407b-b0df-ebf04e41e62f.mp4

Could you please check it ?

Thanks!

HanaRebaiQA commented:
Hello @Seb33300 Yes! the `Overwrite templates` is well enabled (please check my firt screen-record)
I specify that the issue appears in the **new_order** email 

Thanks!

Seb33300 commented:
Can you check that you do not have email template generated in your theme folder?
Theme folder has the priority on the core folder.

If your theme has email template folder, you should select your theme in the list instead of core.

marsaldev commented:
> Hello @marsaldev, maybe you can also add the same changes for the `recycled packaging` option which has the exact same issue.
> 
Hi @Seb33300 
Sure, is there already an issue opened?

Hi @HanaRebaiQA 
I'll make a check asap 😉 but this PR is about the _order_conf_ template, the e-mail read by the customer and not by the merchant 😊  so I'd say that is normal to still have the gift wrapping line in the _new_order_ e-mail 😊

Seb33300 commented:
> Hi @Seb33300
Sure, is there already an issue opened?

I don't think there is a specific issue for the recycled package option.
I did not put it in [my issue](https://github.com/PrestaShop/PrestaShop/issues/22825) because this option did not exist when I created it.

Now that I have upgraded my project to 1.7.8, I am facing it and for me, the issue I described for gift wrapping is the same for recycled package.

marsaldev commented:
Hi @HanaRebaiQA, 
just checked, but as I said, this PR is for `order_conf` e-mail, and not the e-mail received to merchant, sent from `ps_emailalerts` module 😊

atomiix commented on src/Adapter/MailTemplate/MailTemplateTwigRenderer.php:
```diff
@@ -67,12 +70,14 @@ class MailTemplateTwigRenderer implements MailTemplateRendererInterface
         LayoutVariablesBuilderInterface $variablesBuilder,
-        HookDispatcherInterface $hookDispatcher
+        HookDispatcherInterface $hookDispatcher,
+        bool $hasGiftWrapping
```
Can you update the phpdoc?

marsaldev replied:
Hi @atomiix,
if you're referring to the BC I've already opened a PR: https://github.com/PrestaShop/docs/pull/1333 😊

atomiix replied:
No sorry, I was talking about lines 64-66 😅:
```php
/**
 * @param Environment $twig
 * @param LayoutVariablesBuilderInterface $variablesBuilder
 * @param HookDispatcherInterface $hookDispatcher
```

marsaldev replied:
Done 😉

Commit:
phpdoc update

HanaRebaiQA reviewed (commented):
Hello @marsaldev 

Thanks for your feedback!

I have checked again this PR. I still have the issue with `order_conf `when use the `classic` email theme.

Here the steps to reproduce : 

1. BO > CONFIGURE > Shop Parameters > Order Settings
2. Scroll down to Gift options block
3. Disable the `Offer gift wrapping` option and save
4. BO > IMPROVE > Design > Email Theme > Configuration > Select `classic` as the default email theme
5. In Generate emails > Select `classic` , Select the default language , Enable the `Overwrite templates `
6. Click on generate emails
7. In Email themes > Click to Preview the classic theme
8. Search for the `order_conf` and Preview the template with Http 
9. See error > The ` Gift-wrapping`  still displayed
10. Go to FO > Create an order
11. Open the Order Confirmation mail 
12. See error > The ` Gift-wrapping` still displayed

When choose the `modern` email theme, the` Gift-wrapping` is not displayed in the email confirmation 

![image](https://user-images.githubusercontent.com/92912932/164024620-6bd35044-5eb4-4650-8af6-6dd606786419.png)

https://user-images.githubusercontent.com/92912932/164027483-be81f287-a30b-4c6e-9ac8-616628912b41.mp4



Could you please check it ?

Thanks!

marsaldev commented:
Hi @HanaRebaiQA,
thanks a lot for your testing! 
I think it's "normal" because the edits are only on _modern_ template ^^"
I'm adding the fix also for the _classic_ with a new commit 😉

Commit:
Add check also for Classic email template

HanaRebaiQA reviewed (commented):
Hello @marsaldev 

I have checked again this PR and the issue with `classic` theme that i mentioned [here](https://github.com/PrestaShop/PrestaShop/pull/27890#pullrequestreview-945816281) is not fixed.

![image](https://user-images.githubusercontent.com/92912932/165528444-d636cc80-3ef5-4edf-ae56-0d490971c5d7.png)


https://user-images.githubusercontent.com/92912932/165527932-5a900e6e-ffd7-4dac-828f-1e98b6f42182.mp4


Could you please check again ?

Thanks!

marsaldev commented:
Hi @HanaRebaiQA, thanks for your testing!
Sorry, but I'm not able to reproduce your behaviour, see my recording where I enabled and disabled two times the gift wrapping 😊 
(without errors :()

https://user-images.githubusercontent.com/48283112/165726271-42c20d55-e864-41ab-9c89-948254ae07da.mp4

kpodemski commented:
Hello @marsaldev @HanaRebaiQA 

Maybe pair-coding, pair-testing session would be a good idea to proceed with this PR 👍

HanaRebaiQA reviewed (commented):
Hello @marsaldev 

Thanks for your feedback. I also have checked again and i still have the issue with `classic`  email theme.

Thanks @kpodemski and Im OK for check with @marsaldev 

https://user-images.githubusercontent.com/92912932/170501149-25722827-c89b-4e69-8f03-a9876a8b47e1.mp4


Thank you!

Seb33300 commented on mails/themes/classic/core/order_conf.html.twig:
```diff
@@ -146,6 +146,7 @@
             </table>
           </td>
         </tr>
+        {% if giftWrapping == 1 %}
```
You put the condition on the shipping line.

marsaldev replied:
Oh my, you're right!

Seb33300 commented:
@marsaldev you made a mistake on the classic template. The sipping cost line is hidden instead of the gift wrapping line.

Commit:
Fix if condition

HanaRebaiQA reviewed (approved):
Hello @marsaldev 

Thanks for the fixed issue. I have checked again this PR and here is my tested cases : 

-Automated tests : OK 
-Check when `Gift-wrapping` is enabled AND when it's disabled with all combinations of `classic` and `modern` email-theme : OK
-Check when create an order from BO : OK 
-Check when create an order from FO : OK 
-Check email-theme generated from Http from the preview theme list of `modern` and `classic` : OK  
-Check with multistore : Ok 
-Check with different langauges : OK
-Check with `maildev` server : OK  

**Modern email-theme when Gift-wrapping is disabled** :heavy_check_mark: 
![image](https://user-images.githubusercontent.com/92912932/173374077-db3bb069-bfb4-4948-8fc9-1aa733ff0ef3.png)

**Classic email-theme when Gift-wrapping is disabled** :heavy_check_mark: 
![image](https://user-images.githubusercontent.com/92912932/173374213-b07ba913-5d50-49b2-896f-d29b53c1979d.png)


So, it's QA :heavy_check_mark: 

Thank you!

Progi1984 commented:
Thanks @marsaldev & @HanaRebaiQA

Progi1984 commented:
@PrestaShop/prestashop-maintainers Is this PR needs to be reverted as it brings "BC Breaks"?

Progi1984 commented:
@PrestaShop/prestashop-core-developers  Is this PR needs to be reverted as it brings "BC Breaks"?

kpodemski commented:
@Progi1984 let's discuss this on Tuesday during a meeting

Progi1984 commented:
@kpodemski Please tell me the decision you took after this one.
